### ISSUE TITLE
Fatal error in backend; zero extension from i8 not supported?

### ISSUE THREAD

**Issue by Blei** [llvm-dcpu16/llvm-dcpu16#81]
The following short program:

``` c
int hailstone(int x) {
  int counter = 0;
  while (x!=1) {
    counter++;
    if (x&1) {
      x = x * 3 + 1;
    } else {
      x = x/2;
    }
  }
  return counter;
}

int main() {
  return hailstone(3);
}
```

fails with this error:

```
fatal error: error in backend: Cannot select: 0x5c2a540: i16,ch = load 0x5c2a240, 0x5c2a640,
      0x5c2a340<LD1[%x.addr](align=2), zext from i8> [ID=11]
  0x5c2a640: i16 = FrameIndex<0> [ORD=9] [ID=4]
  0x5c2a340: i16 = undef [ORD=6] [ID=2]
```

**Comment by asl** [comment]
The problem is with copying the code from MSP430.
Adding something like

```
setLoadExtAction(ISD::ZEXTLOAD, MVT::i8,  Expand);
```

into ctor of DCPU16TargetLowering should fix the issue.

**Comment by ghost** [comment]
```
setLoadExtAction(ISD::ZEXTLOAD, MVT::i8,  Expand);
```

Doesn't do the trick. He want's ANY_EXTEND set then. If we do that, we get a nice assertion:

```
Assertion `ExtType < ISD::LAST_LOADEXT_TYPE && VT < MVT::LAST_VALUETYPE 
    && "Table isn't big enough!"' failed.
```

**Comment by ghost** [comment]
Very odd. I really ask myself why he screams for a ANY_EXTEND for i8. When I look at the LLVM IR, I can't see a i8 at that position of the code:

```
  %and1 = and i16 %4, 1
  store i16 %and1, i16* %x.addr, align 2
  br label %if.end
```

**Comment by asl** [comment]
> Doesn't do the trick. He want's ANY_EXTEND set then. If we do that, we get a nice assertion:

What's about Promote then?

> When I look at the LLVM IR, I can't see a i8 at that position of the code

It's inserted by DAG combiner. First, it proves that %and1 is really i1 and thus it creates zextload / truncstore of i1. However, i1 is marked as "Promote", so it's promoted to next legal type - i8 here. So, I'd suggest to remove all expansions of i8 {S,Z}EXTLOADS and make sure i1 is promoted to i16.

**Comment by ghost** [comment]
Okay, I see. Promote didn't bring the trick. Right now I tried to change the Datalayout to

```
DataLayout("e-p:16:16:16-i16:16:16-i32:16:32-n16")
```

Since it included native i8. Then I changed in the DCPU16ISelLowering.cpp:

```
  setLoadExtAction(ISD::EXTLOAD,  MVT::i1,  Promote);
  setLoadExtAction(ISD::SEXTLOAD, MVT::i1,  Promote);
  setLoadExtAction(ISD::ZEXTLOAD, MVT::i1,  Promote);
  setLoadExtAction(ISD::ZEXTLOAD, MVT::i8,  Promote);
  setLoadExtAction(ISD::SEXTLOAD, MVT::i16, Expand);
```

To promote the i1 to (hopefully) i16. When I exclude the promote for i8, he wants a ZEXTLOAD expansion (which won't work). But then I will get a 'Action not implemented' error:

```
This action is not supported yet!
UNREACHABLE executed at ~/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:1087!
```

Is there any way to to get more information which action is not supported? I already had those problems often and find them hard to debug.

**Comment by asl** [comment]
Well, the only way which you can trust is reading the code in LegalizeDAG. I'd suggest removing all loadext actions and start from simple promote for all i1 extloads. And then - check what's missing.

**Comment by ghost** [comment]
Okay, I think the best way to resolve this, is to set the following in the DCPU16ISelLowering.cpp:

```
setLoadExtAction(ISD::EXTLOAD,  MVT::i8,  Promote);
setLoadExtAction(ISD::SEXTLOAD, MVT::i8,  Promote);
setLoadExtAction(ISD::ZEXTLOAD, MVT::i8,  Promote);
```

This promotes every i8 to i16. Then we "just" have to know which action is not supported by us.

**Comment by ghost** [comment]
Just for reference:

Builds with the changes I suggested above:

```
; ModuleID = 'good.c'
target datalayout = "e-p:16:16:16-i8:8:8-i16:16:16-i32:16:32-i64:64:64-n8:16"
target triple = "dcpu16"

define i16 @ncmp(i16 %x) nounwind {
entry:
  %retval = alloca i16, align 2
  %x.addr = alloca i16, align 2
  store i16 %x, i16* %x.addr, align 2
  %0 = load i16* %x.addr, align 2
  %cmp = icmp ne i16 %0, 0
  br i1 %cmp, label %if.then, label %if.end

if.then:                                          ; preds = %entry
  %1 = load i16* %x.addr, align 2
  store i16 %1, i16* %retval
  br label %return

if.end:                                           ; preds = %entry
  store i16 0, i16* %retval
  br label %return

return:                                           ; preds = %if.end, %if.then
  %2 = load i16* %retval
  ret i16 %2
}
```

Won't build:

```
; ModuleID = 'bad.c'
target datalayout = "e-p:16:16:16-i8:8:8-i16:16:16-i32:16:32-i64:64:64-n8:16"
target triple = "dcpu16"

define i16 @hailstone() nounwind {
entry:
  %retval = alloca i16, align 2
  %x = alloca i16, align 2
  store i16 0, i16* %x, align 2
  %0 = load i16* %x, align 2
  %and = and i16 %0, 1
  %tobool = icmp ne i16 %and, 0
  br i1 %tobool, label %if.then, label %if.else

if.then:                                          ; preds = %entry
  store i16 0, i16* %retval
  br label %return

if.else:                                          ; preds = %entry
  store i16 1, i16* %retval
  br label %return

return:                                           ; preds = %if.else, %if.then
  %1 = load i16* %retval
  ret i16 %1
}
```

**Comment by ghost** [comment]
Okay, it seems to be a problem with the AddrMode and loading. Explanation:

Works:

```
int k = x&1;
if(x&1) { }
```

Won't work:

```
if(x&1) { }
```

If you use 'llc --march=dcpu16 test.llvm -o test.s -view-legalize-dags' , then you will see a call for "load<LD1[%x.addr](align=2), zext from i8>", which will be promoted to i16, but can't be assembled, because such an instruction seems to be missing.

I will investigate more.

**Comment by asl** [comment]
How it will be promoted? You see, this is not real load, it's zextload from i8.

**Comment by ghost** [comment]
_mhh_ You're right. Now the questions should only be: How do we promote it?

```
setOperationAction(ISD::ZEXTLOAD,          MVT::i8,   Promote);
```

Didn't do the trick.

BTW: I feel like a bloody trainee.

**Comment by asl** [comment]
Right, you cannot Promote further, because then extload won't be extload :) So, you have either custom lower it (e.g. via loading i16 and trunc'ing it) or find some other way how you can load just i8 value...

**Comment by ghost** [comment]
I wonder why he insist on using a i8 here. Shouldn't LLVM generate a DAG with the SelectionDAG LegalizeTypes Phase which only contains legal/valid types?

**Comment by asl** [comment]
It's hard to tell right now w/o digging into the code... It's thinking that extload from i8 to i16 is somehow legal...

**Comment by ghost** [comment]
Okay, so my suspicion was right. I will look deeper into that lead.

**Comment by ghost** [comment]
Okay, found a solution.

**Comment by ghost** [comment]
Fixed with commit 1dce60768071be0cbc15bcc04b0ae95b145a640b

**Comment by asl** [comment]
The "fix" does not seem to be correct
